Several improvements and bugfixes to cocos2d

rsanchezsaez opened the pull request:
Dear Riq,

I'm sending you a pull request with some bugfixes and improvements to Cocos2D.

Summary of main changes:
- Fix for touch point conversion bugs in the CCNode's convertTouchToNodeSpace: and CCMenu's itemForTouch: methods. This issue goes undetected if only one OpenGLView is used (as most of the folks do), but if you add overlaid subviews to the OpenGLView (like an UIScrollView), then touches are not translated properly into cocos. In my opinion this is the most important improvement.
- Added new director type (kCCDirectorTypeDisplayLinkUIKit) that behaves correctly with UIKit Objects. t's exactly like kCCDirectorTypeDisplayLink with just two differences: a) uses 'NSRunLoopCommonModes' when attaching DisplayLink to CurrentLoop  to prevent animation freezing when holding a still touch on the scrollview; b) throttles framerate in mainLoop  to 90fps (needed when UIScrollView takes control in scrolling or zooming (UITrackingRunLoopMode) or when [setContentOffset:offset animated:YES] is called; otherwise OpenGL over-rendering breaks UIScrollView inertial behaviour).
- Added CCLayerRGBA subclass, which implements the CCRGBAProtocol. It propagates the 'opacity' and 'color' values to the children that conform to the CCRGBAProtocol.
- Made CCMenu opacity and color properties writable.

Cheers and keep up the good work with cocos2d!

rsanchezsaez committed:
CCNode: Fixed incorrect point translation from touched view space to openGLView space when the touched view is not the openGLView (makes cocos2d workable with an overlaid UIScrollView).

rsanchezsaez committed:
ES1Renderer: Commented CCLOG in -(BOOL)resizeFromLayer:(CAEAGLLayer*)layer method because using a UIScrollView as an OpenGLView overlaid subview causes this method to be called continuously when scrolling.

rsanchezsaez committed:
CCMenu: Fixed incorrect point translation from touched view space to openGLView space when the touched view is not the openGLView (makes cocos2d workable with an overlaid UIScrollView).

rsanchezsaez committed:
CCMenu and CCMenuItem: Made 'opacity' and 'color' properties writeable.

rsanchezsaez committed:
CCDirectorTimer: Uncommented line to made this director (kCCDirectorTypeNSTimer) compatible with UIScrollView and other UIKit objects.

rsanchezsaez committed:
CCDirector and CCDirectorIOS: Implemented new director: CCDisplayLinkDirectorUIKit. It is a Director that synchronizes timers with the refresh rate of the display and works correctly with UIKit objects (like UIScrollView).

It's exactly like kCCDirectorTypeDisplayLink with just two differences:
 -  Using 'NSRunLoopCommonModes' when attaching DisplayLink to CurrentLoop  to prevent animation freezing
    when holding a still touch on the scrollview.
 -  Throttling framerate in mainLoop  to 90fps (needed when UIScrollView takes control in
    scrolling or zooming (UITrackingRunLoopMode) or when [setContentOffset:offset animated:YES] is called;
    otherwise OpenGL over-rendering breaks UIScrollView behaviour).

rsanchezsaez committed:
CCLayer: Implemented CCLayerRGBA subclass, it allows to recursively set opacity and color of Layer's children.

rsanchezsaez committed:
Merge commit 'release-1.0.0-rc2' into rsanchez

ricardoquesada commented:
Thanks Ricardo.
Si es posible, podras agregar un test para el nuevo Director. Estoy interesado es ver un tests para endenter un poco mejor este nuevo director. Y ya que es muy parecido al DisplayLink director, quizas tenga sentido merger el nuevo Director y el DisplayLinkDirector en uno solo.

Gracias.

rsanchezsaez commented:
Hola Ricardo.

Cuando tenga tiempo miraré lo de agregar un test para el nuevo Director.

Realmente DisplayLinkUIKit es una clase heredada de DisplayLink director, en la que solo se han modificado los metodos:
-(void)startAnimation, para añadir el displayLink al currentLoop en modo NSRunLoopCommonModes
-(void)mainLoop:(id)sender, para ponerle freno a la llamada drawScene, y hacer que el dibujador de OpenGL no sea llamado continuamente cuando estas utilizando un UIScrollView. Lo he hecho así:

(void)mainLoop:(id)sender
{    if ([self getDeltaTimePre]>1.0/90) {
        [self drawScene];    }
}

Este metodo usa un nuevo método, -(float)getDeltaTimePre, que es igual que -(void)calculateDeltaTime, pero en vez de actualizar el dt_ interno, lo calcula y lo devuelve sin actualizarlo.

Si no se le pone límite al mainLoop, en apariencia todo funciona bien, peor al empezar a usar cocos2d con un UIScrollView y hacer scroll sobre él, el UIScrollView funciona erráticamente (por ejemplo, deja de tener inercia, el metodo setCurrentOffset:animated:YES no va bien, etc.) porque cuando el UIScrollView está en operación, se llama a la función mainLoop: continuamente (en vez de 60fps, como es lo normal para el displayLink) y ahoga al hilo que procesa los touches del UIScrollView.

A lo mejor hay alguna solución más elegante, pero estuve investigando mucho y no se me ocurrió nada mejor.

Un saludo,
Ricardo

rsanchezsaez commented:
Hola,

estoy mirando los tests. ¿Hay algún test que recorra los 4 directores y mida el rendimiento de cada director? No encuentro ningún test así. ¿Como debo implementar el test del nuevo director, en cuál me baso?

Un saludo,
Ricardo

ricardoquesada commented:
Hola Ricardo,

La verdad que no hay ningun test que pruebe los directores. En los
tiempos del iOS 2.x yo tenia una tests muy ad-hoc, pero creo que no lo
comitié (y no lo encuentro).

Si podes, crearía un nuevo tests que pruebe tu director. Creo que el
test "Attach Test" es un buen ejemplo para empezar ya que usa objectos
UIKit (aunque creo que no los usa dentro del EAGLView), y crea y
destruye directores.

Gracias!

2011/5/29 rsanchezsaez
reply@reply.github.com:

> Hola,
> 
> estoy mirando los tests. ¿Hay algún test que recorra los 4 directores y mida el rendimiento de cada director? No encuentro ningún test así. ¿Como debo implementar el test del nuevo director, en cuál me baso?
> 
> Un saludo,
> Ricardo
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/cocos2d/cocos2d-iphone/pull/60#issuecomment-1259563

rsanchezsaez committed:
CCArray: Fixed warning about unsigned int losing precision when compiling cocos2d-mac projects for 64bits.

rsanchezsaez committed:
CCFileUtils: Fixed possible memory leak in due to usage of deprecated method.

rsanchezsaez committed:
CCNode: Added -addChild:tag: method.

rsanchezsaez committed:
CCDirector: Implemented customizable position of FPS label.

rsanchezsaez committed:
CCLayer: Added isEventEnabled (Event = Touch, Mouse, Keyboard & Accelerometer)  propagation to children supporting input events (CCLayer and CCMenu mainly).

rsanchezsaez committed:
CCDirector: Fixed bug in custom FPS label positioning.

rsanchezsaez committed:
CCSpriteFrameCache: Adding frames to the cache now returns the added frames in an NSArray. Manually added frames should be retained by the CCLayer or CCSprite using them, so they don't disappear on cache puring.

rsanchezsaez committed:
CCFileUtils: Added 'CC_IPAD_UNIVERSAL_SUPPORT' (Using it on a Universal app running on iPad will try to load assets with the '-ip' suffix first, then '-hd' asset, then regular assets).

rsanchezsaez committed:
SimpleAudioEngine: Added backgroundMusicCurrentTime read-write property.

rsanchezsaez committed:
CocosDenshion: Fixed NSString over-reatining.

rsanchezsaez committed:
CCDirector/CCDirectorIOS: Improved some methods so popScene doesn't break CCDirector after popping the last scene (so runWithScene can be used to resume director again).

rsanchezsaez committed:
CCDirector/CCDirectorIOS: Performance improvements in calculateDeltaTime. Using the Apple recommended CACurrentMediaTime now.

rsanchezsaez committed:
OpenGL drawing code optimization: replaced glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT) by glClear(GL_COLOR_BUFFER_BIT ). Works fine if you only use 2D graphics.

rsanchezsaez committed:
glu.c: Added optimizations from http://www.cocos2d-iphone.org/forum/topic/10433

rsanchezsaez committed:
CCactionInterval: Added CCAnimateAt action (allows animating a sprite starting at an arbitrary (or random) frame). actionWithRandomStartAnimation can be used for adding several idle animated sprites, so they animate synchronously.

rsanchezsaez committed:
CCDirector: Added popSceneWithCleanup:(BOOL)cleanup so persistent (retained) scenes can be popped without performing cleanup.

rsanchezsaez committed:
CCNode: Added removeFromParent method, so it can be used with CCCallFunc actions.

rsanchezsaez committed:
CCDirectorIOS: Added some commented CCDirectorDisplayLinkUIKit testing code.

rsanchezsaez committed:
CCScheduler: Fixed purgeSharedScheduler to allow purging the shared scheduler within a scheduled method.

rsanchezsaez committed:
Various: Updated some documentation comments regarding some of my changes.

rsanchezsaez committed:
CCDirectorMac: Fixed bug regarding optimized time calculation.

rsanchezsaez committed:
CCFileUtils, ccConfig: Fixed bug preventing Mac compilation due to CC_IPAD_UNIVERSAL_SUPPORT (-ip extension).

rsanchezsaez committed:
CCActionInterval: Added CCMoveToEx and CCMoveByEx actions. These can be called concurrectly on the same node (the resulting ovement will the addition of all the called movements).

rsanchezsaez committed:
CCActionManager: Minor code improvements on init method.

rsanchezsaez committed:
CCScheduler: Rolling back, removed delayed SharedScheduler purge.

rsanchezsaez committed:
CCSprite: Changed color of CC_SPRITE_DEBUG_DRAW  bounding boxes for them to be more visible.

rsanchezsaez committed:
CCLabelTTF, CCTexture2D: Added vertical alignment option when rendering strings to textures.

rsanchezsaez committed:
CCParticleSystem: Made particle plist loading to be path independent (it looks for texture image file on the same path as the loaded plist).

rsanchezsaez committed:
CCDirector: Implemented popSceneClass and pushSceneClass. This methods allows for complete deallocation of current scene before beginning to allocate the new one.

You must pass [yourScene class] as the argument. The class must conform to the CCSceneAutoreleased protocol (it must have an argument-less +(id)scene method which allocates and returns the autorelease scene).

The old pushScene and popScene work as usual, but the code has been improved a bit:
- popping the last scene no longer calls [director end]. Instead, just the caches are purged.
- popping the last scene won't purge the caches if you are pushing a scene that has been alread initialized
  (because puring the caches would destroy data owned by the new scene)
- pushScene: now acts as runWithScene:, which is now deprecated) (the only difference is that you must
  manually call [director startAnimation] at the end of applicationDidFinishLaunching:)
- CCDirector end and dealloc methods have been sanitized a little bit to make it more Cocoa like. Some objects are deallocated at CCDirector's -(void)dealloc method, rather than at -(void)end.

rsanchezsaez committed:
CCLabelTTF, CCTexture2D: Fixed Mac compilation of newly introduced vertical alignment.

rsanchezsaez committed:
CCActionInterval: Fixed Mac compilation of CCMoveByEx.

rsanchezsaez committed:
cocos2d, CCLabelTTF: Added fontName, fontSize and string properties.

rsanchezsaez committed:
Merge commit 'release-1.0.1' into rsanchez. Performed manual merge on conflicting files keeping all the rsanchez-1.0.0 features and optimizations.

Conflicts:
	cocos2d/CCActionInterval.h
	cocos2d/CCActionInterval.m
	cocos2d/CCDirector.m
	cocos2d/CCLabelTTF.h
	cocos2d/CCLabelTTF.m
	cocos2d/CCProtocols.h
	cocos2d/CCSpriteFrameCache.m
	cocos2d/CCTexture2D.m
	cocos2d/Platforms/iOS/CCDirectorIOS.m

rsanchezsaez committed:
Fixed some bugs regarding Mac compilation introduced by the 1.0.1 merge.

rsanchezsaez committed:
CCDirector: Improved new pushScene and popScene system. Now runScene, although deprecated, works properly. You no longer have to call sharedDirector's animationStart manually; pushScene/runScene takes care of calling it at the correct time the first time.

rsanchezsaez committed:
CCNode, CCSprite: 1) Fixed 'disable subpixel rendering' option. 2) Improved it to use the round() function (rather than (int)cast).

rsanchezsaez committed:
CGpointExtension: Added ccpRound() function.

rsanchezsaez committed:
Implemented propagated opacity and added CC_PROPAGATE_OPACITY option in ccConfig.h.

If enabled, cocos2d uses transitive opacity. All classes calculate its cumulative opacity from self opacity and parent's opacity. For example, if you setup a CCLayerRGB with an opacity of 127 (50%) and add a children CCSprite with an opacity of 127,
it draws like having an opacity of 63 (25%).

For performance reasons, it's recommended to disable it if you don't need the feature.
To enable set it to 1. Use 0 to disable it. Enabled by default

rsanchezsaez committed:
CCLabelBMFont: Fixed small bug in propagated opacity.

rsanchezsaez committed:
SpriteTest: Added SpriteTestDelay animation (for testing unfixed issue #1020).

rsanchezsaez committed:
CCNode: Improved removeFromParent method to make it behave nicer on subclassing.

rsanchezsaez committed:
CCDirectorIOS: Removed CCDirectorDisplayLinkUIKit (it was not a correct/efficient solution, let's hope Apple fixes the CADisplayLink bug (that makes CCDirectorDisplayLink not behave properly with UIKit elements) on iOS 5.

rsanchezsaez committed:
CCNode and CCActionManager: Added CCNode's -(void)stopAllActionsByTag:(NSInteger)tag and CCActionManager's -(void)removeAllActionsByTag:(NSInteger)tag target:(id)target methods.

rsanchezsaez committed:
CCActionManager, CCScheduler: Commented out debug messages.

rsanchezsaez committed:
CCParticleSystem: Added startSystem method for restarting a stopped system (caution, do not use if you have autoRemoveOnFinish enabled!).

rsanchezsaez committed:
ccTypes.h: Added ccc4F convenince method and improved ccc3B and ccc4B convenince methods.

rsanchezsaez committed:
CCActionInterval: Small optimization to CCAnimate and CCAnimateAt update methods.

rsanchezsaez committed:
CCActionInterval: Added actionWithRandomStartAnimation:factor: method to CCAnimateAt. It allows creating factored random start animation (animation initial frame is randomized within the given factor). This, used with low factors, allows for subtle animation unsynchronization between sprites performing the same animation at the same time.

rsanchezsaez committed:
CActionInterval: Added nil checks to CCAnimateAt contructors.

rsanchezsaez committed:
CCNode: Improved onEnter base method to allow inserting nodes on self.parent during a subclassed onEnter.

rsanchezsaez committed:
Minor changes that solve some Xcode Analyzer warnings.

rsanchezsaez committed:
CCRibbon: Fixed some warnings when compiling on Mac OS X 64bit.

rsanchezsaez committed:
CCActionInstant: Added CCSetValue action, which allows to set any numeric object member to a given value.

rsanchezsaez committed:
CCmenu: Minor optimization.

rsanchezsaez committed:
CCParticleSystem: Added new kCCPositionTypeRelativeToGrandparent emitter type (in this mode, particles are 'attached' to the grandparent layer, useful for drawing particles on a scrollable map).

rsanchezsaez committed:
1) Updated cocos to 1.1-beta2
2) Fixed bug in CCRepeat action

Merge commit 'release-1.1-beta2' into rsanchez

Conflicts:
	cocos2d-ios.xcodeproj/project.pbxproj
	cocos2d/CCDirector.m
	cocos2d/CCParticleSystem.m
	cocos2d/CCRibbon.m
	cocos2d/CCTexture2D.h
	cocos2d/Support/CCFileUtils.m
	cocos2d/ccConfig.h
	tests/SpriteTest.h
	tests/SpriteTest.m

rsanchezsaez committed:
CCLabelTTF: Added setFontSize method.

rsanchezsaez committed:
CCLayer: Disabled isTouchEnabled propagation.

rsanchezsaez committed:
CCNode: Improved onExit so you can use it to remove sibling CCNodes (parent's children).

rsanchezsaez committed:
CCParticleSystem: Fixed kCCPositionTypeRelativeToGrandparent mode so it supports grandparent scaling.

rsanchezsaez committed:
CCSprite: Improved CC_SPRITE_DEBUG_DRAW mode so it draws to diagonal lines crossing the sprite box.

rsanchezsaez committed:
CCSpriteBatchNode: Allowing texture less sprites (containers for further sprites).

rsanchezsaez committed:
CCSpriteFrame: Added originalSize property.

rsanchezsaez committed:
CCTexture2D: Fixed a text-texture size bug in retina.

rsanchezsaez committed:
CCFileUtils: Improved iPad asset reading code so it falls back to retina assets.

rsanchezsaez committed:
Re-enabled subpixel rendering.

rsanchezsaez committed:
ccMacros: Added new CC_SIZE_PIXELS_TO_POINTS and CC_SIZE_POINTS_TO_PIXELS macros.

rsanchezsaez committed:
CCActionInstant: Updated CCCallBlock action so it does nothing for an nil block (rather than crashing).

rsanchezsaez committed:
CCNode, CCSprite:
1) Added moveBy method (updates position with a position delta, rather than an absolute position; used by *Move*Ex actions).
2) Added unanchoredPosition method (which allows to obtain the center of the sprite regardless of its AnchorPoint).

rsanchezsaez committed:
CCActionInterval: Updated CCMove*Ex actions to use new [CCNode moveBy] method.

rsanchezsaez committed:
CCMenuItem: Aesthetic improvements to CCMenuItemLabel touch effect.

rsanchezsaez committed:
CCNode: Added -(CGPoint)anchoredPosition:(CGPoint)anchorPoint method. Returns the position modified by the specified anchor point (without updating the anchorPoint member variable).

rsanchezsaez committed:
FontLabelStringDrawing: Added '¿' and '¡' to list of non-line-wrapping characters (needed for Spanish).

rsanchezsaez committed:
CCTexture2D: Use zFont as default font type (it uses custom line-wrapping that can be customized for further languages).

rsanchezsaez committed:
CCParticleSystem: Fixed bug in kCCPositionTypeRelativeToGrandparent type and added warning about scaling using this type of emitter.

rsanchezsaez committed:
CCNode: Added -(int)zOrderOfHighestChildren method.

rsanchezsaez committed:
cocos2d, CCActionInterval: Solved bug in CCAnimate introduced by previous intented optimization.

rsanchezsaez committed:
ccTypes: Added ccc3BEqual(ccColor3B a, ccColor3B b) method.

rsanchezsaez committed:
CCActionInterval: Added [CCTintTo actionWithDuration:(ccTime)t color:(ccColor3B)color] convenience constructor.

rsanchezsaez committed:
CCSprite: Added "-(void)setDisplayFrameWithName:(NSString*)frameName" method.

rsanchezsaez committed:
CCSpriteFrameCache: Added "-(CCSpriteFrame*) addSpriteFrame:(CCSpriteFrame*)frame name:(NSString*)frameName" and "-(CCSpriteFrame*) addSpriteFrameWithTextureFile:(NSString*)textureFilename".

rsanchezsaez committed:
CCNode: Added -(CGPoint)centerPoint method, which returns the node's content size center point.

rsanchezsaez committed:
CCSpriteFrameCache: Added "-(CCSpriteFrame*) addSpriteFrameWithTextureFile:(NSString*)textureFilename name:(NSString*)name" method. Changed behavior of "-(CCSpriteFrame*) addSpriteFrameWithTextureFile:(NSString*)textureFilename" method, now the SpriteFrame name is uses just the texture name (without the full path) with '.png' extension.

rsanchezsaez committed:
CCMenu: Overriden -(void)setIsTouchEnabled:(BOOL)isTouchEnabled so menu status is reset before enabling/disabling touch. This fixes a problem in which CCMenu stops working if you disable touches while a button is pressed.

rsanchezsaez committed:
cocos2d, CCArray: Added CCARRAY_FOREACH_INVERSE convenience macro for iterating CCArrays in reverse order.

rsanchezsaez committed:
cocos2d, CCMenu: Touched items now are processed in the inverse order (so overlapping items are processed correctly).

rsanchezsaez committed:
cocos2D, CCNode: Solved a possibly harmful bug in orderOfArrival synthesized property.

rsanchezsaez committed:
CCNode, Fixed small bug in zOrderOfHighestChildren method.

rsanchezsaez committed:
Merge commit 'release-1.1-beta2b' into rsanchez

Conflicts:
	cocos2d-ios.xcodeproj/project.pbxproj
	cocos2d/CCActionInterval.h
	cocos2d/CCActionInterval.m
	cocos2d/CCAtlasNode.m
	cocos2d/CCLabelBMFont.m
	cocos2d/CCNode.h
	cocos2d/CCSprite.m
	cocos2d/CCTexture2D.m
	cocos2d/Platforms/iOS/CCDirectorIOS.h
	cocos2d/Platforms/iOS/ES1Renderer.m
	cocos2d/Support/CCFileUtils.m
	tests/SpriteTest.h

rsanchezsaez committed:
CCActionInterval: Fixing CCAnimateAt class for new CCAnimation version.

rsanchezsaez committed:
CCAnimation: Explicitly initializing new CCAnimation members.

rsanchezsaez committed:
CCActionInterval: Reverting pseudo-fix that was correctly fixed elsewhere.

rsanchezsaez committed:
CCActionInterval: Removing temporal debug code.

araker commented:
I've added the fixes to the menu (changing opacity and color) and replaced [touch view] with [Director OpenGLView] view]. 

The new director type looks good for the UIScrollView problem. I like to have one solution that tackles all problems, I'll try this one out. 

The other commits are already implemented, or not relevant anymore.

araker commented:
Closing this one as well, the director now has a selectable run loop, effectively the same as the new director type.
